File: data structures/Library Query.py
# -*- coding: utf-8 -*-
"""
Created on Mon Aug 31 12:53:34 2020

@author: rahul
"""


class BinaryIndexedTree2D:
    '''1 based indexing'''

    def __init__(self, max_x=10000):
        self.x_max = 10000
        self.y_max = 1000
        self.tree = [[0 for i in range(self.y_max + 5)] for j in range(self.x_max + 5)]

    def update(self, x, y, val):
        while x <= self.x_max:
            y1 = y
            while y1 <= self.y_max:
                self.tree[x][y1] += val
                y1 += (y1 & -y1)
            x += (x & -x)

    def getFrequency(self, x, y):
        sum_ = 0
        while x > 0:
            y1 = y
            while y1 > 0:
                sum_ += self.tree[x][y1]
                y1 -= (y1 & -y1)
            x -= (x & -x)
        return sum_


def solve(t, l, r, k):
    low = 1
    high = t.y_max
    ans = -1
    while low <= high:
        mid = (low + high) // 2
        v2 = t.getFrequency(r, mid)
        v1 = t.getFrequency(l - 1, mid)
        if v2 - v1 >= k:
            ans = mid
            high = mid - 1
        else:
            low = mid + 1
    return ans


s = ''

for _ in range(int(input())):
    n = int(input())
    tree = BinaryIndexedTree2D(n)

    no_books = [0] + list(map(int, input().split()))
    for i in range(1, n + 1):
        tree.update(i, no_books[i], 1)

    q = int(input())
    for i in range(q):
        query = list(map(int, input().split()))
        t = query[0]
        if t == 0:
            x, y, k = query[1:]
            res = solve(tree, x, y, k)
            s += str(res) + '\n'
        else:
            x, k = query[1:]
            tree.update(x, no_books[x], -1)
            no_books[x] = k
            tree.update(x, no_books[x], 1)

print(s)

File: data structures/Library Query.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;


int main() {
    int t;
    cin >> t;
    for (int e = 0; e < t; e++) {
        int n, q, t, x, y, k, i, l;
        cin >> n;
        int no_books[n + 1];
        no_books[0] = 0;
        for (i = 1; i <= n; i++) {
            cin >> no_books[i];
        }
        cin >> q;
        while (q--) {
            cin >> t;
            if (t == 0) {
                cin >> x >> y >> k;
                int temp[1001];
                for (int j = 1; j < 1001; j++) {
                    temp[j] = 0;
                }
                for (int j = x; j <= y; j++) {
                    temp[no_books[j]]++;
                }
                int sum = 0, ans = 0;
                for (int j = 1; j <= 1000; j++) {
                    sum += temp[j];
                    if (sum >= k) {
                        ans = j;
                        break;
                    }
                }
                cout << ans << endl;
                for (int j = 1; j < 1001; j++) {
                    temp[j] = 0;
                }

            } else {
                cin >> x >> k;
                no_books[x] = k;
            }
        }
    }
    return 0;
}
